oauth.go

```diff
@@ -0,0 +1,120 @@
+package twilio
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+	"github.com/twilio/twilio-go/client"
+	iam "github.com/twilio/twilio-go/rest/iam/v1"
+)
+
+// TokenAuth handles token-based authentication using OAuth.
+type TokenAuth struct {
+	// token is the cached OAuth token.
+	token string
+	// OAuth is the OAuth client used to fetch new tokens.
+	OAuth client.OAuth
+}
+
+// NewTokenAuth creates a new TokenAuth instance with the provided token and OAuth client.
+func (t *TokenAuth) NewTokenAuth(token string, o client.OAuth) *TokenAuth {
+	return &TokenAuth{token: token, OAuth: o}
+}
+
+// FetchToken retrieves the current token if it is valid, or fetches a new token using the OAuth client.
+func (t *TokenAuth) FetchToken(ctx context.Context) (string, error) {
+	expired, err := t.Expired(ctx)
+	if err != nil {
+		return "", err
+	}
+	if t.token != "" && expired {
+		return t.token, nil
+	}
+
+	token, err := t.OAuth.GetAccessToken(context.Background())
+	if err != nil {
+		return "", err
+	}
+
+	t.token = token
+	return t.token, nil
+}
+
+// Expired checks if the current token is expired.
+func (t *TokenAuth) Expired(ctx context.Context) (bool, error) {
+	token, _, err := new(jwt.Parser).ParseUnverified(t.token, jwt.MapClaims{})
```

Comment:
```suggestion
	token, _, err := new(jwt.Parser).ParseUnverified(t.token, jwt.MapClaims{})
```

Let's not create a new `jwt.Parser` every time - that's a lot of memory being created and then removed. We could just make a package level variable if we need an instance.